### PR TITLE
[tem-1551] Introduces instance stop

### PR DESCRIPTION
**Pull request by mkrisher**
This PR introduces the `instance stop` command to go along with the already merged `instance start` command. Both require a name to be passed and handle the scenario where no config exists for the given name.

Some refactoring was done to move the Docker commands to the Docker object. The `start` command got a little smarter. If a container already exists for the config, start it, if not build one and start it.

### CODE REVIEW

File: src/cmd/instance/create.rs [hunk 1]
```diff
@@ -87,8 +87,6 @@ fn persist_instance_config(matches: &ArgMatches) -> Result<(), Box<dyn Error>> {
     let path = Config::full_path(matches);
     let mut config = Config::new(matches, &path); // calls init and writes the file
 
-    dbg!(&matches);
```
Comment (mkrisher) [inline src/cmd/instance/create.rs:90]: annotation: missed this debug statement in the create command

File: src/cmd/instance/start.rs [hunk 2]
```diff
@@ -40,41 +39,10 @@ pub fn execute(args: &ArgMatches) -> Result<(), Box<dyn Error>> {
...
-                sp.stop_with_message(message);
-
-                let stderr = String::from_utf8(output.stderr).unwrap();
-
-                if !stderr.is_empty() {
-                    return Err(Box::new(DockerError::new(
-                        format!("There was an issue starting the instance: {}", stderr).as_str(),
-                    )));
-                }
+                Docker::start(name, instance)?;
```
Comment (mkrisher) [inline src/cmd/instance/start.rs:45]: annotation: moved the existing start logic to the Docker object

File: src/cli/docker.rs [hunk 3]
```diff
@@ -41,6 +43,122 @@ impl Docker {
 
         Ok(())
     }
+
+    // start container if exists for name otherwise build container and start
+    pub fn start(name: &str, instance: &Instance) -> Result<(), Box<dyn Error>> {
```
Comment (mkrisher) [inline src/cli/docker.rs:48]: annotation: move the existing logic from the subcommand file to this object

File: src/cli/docker.rs [hunk 4]
```diff
@@ -41,6 +43,83 @@ impl Docker {
...
+            .unwrap()
+            .contains("tembo-pg")
+        {
+            println!("- existing container found");
+
+            instance.start();
+        } else {
+            println!("- building and then running container");
+
+            let _ = instance.init();
```
Comment (mkrisher) [inline src/cli/docker.rs:59]: annotation: instance.init will install the extensions the stack defines as the container is built

File: src/cli/extension.rs [hunk 5]
```diff
@@ -0,0 +1,20 @@
+// Extensions are defined in Tembo Stacks and can be custom installed by users
+// they are extensions to Postgres
+
+use crate::{Deserialize, Serialize};
+use chrono::{DateTime, Utc};
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
```
Comment (mkrisher) [inline src/cli/extension.rs:7]: annotation: moved this out of the stack file

File: src/cli/instance.rs [hunk 6]
```diff
@@ -38,3 +45,169 @@ pub struct ExtensionLocation {
...
+        if !stderr.is_empty() {
+            return Err(Box::new(DockerError::new(
+                format!("There was an issue starting the instance: {}", stderr).as_str(),
+            )));
+        }
+
+        Ok(())
+    }
+
+    fn install_extension(&self, extension: &TrunkInstall) -> Result<(), Box<dyn Error>> {
```
Comment (mkrisher) [inline src/cli/instance.rs:142]: annotation: first pass is using trunk to install extensions, will need to be upgraded to support libraries, looking into using the operator logic on next PR

**Review by nhudson — APPROVED** [review]

**Pull request merged**